eth/sync.go

```diff
@@ -223,7 +223,12 @@ func (cs *chainSyncer) loop() {
 
 		case <-cs.pm.quitSync:
 			if cs.doneCh != nil {
-				cs.pm.downloader.Terminate() // Double term is fine, Cancel would block until queue is emptied
+				// Disable all insertion on the blockchain. This needs to happen before
+				// terminating the downloader because the downloader waits for blockchain
+				// inserts to finish and these can take a long time to finish.
+				cs.pm.blockchain.StopInsert()
```

Comment:
Question: what if `doneCh` is nil, and we use the deferretd `downloader.Terminate()` instead of this clause. Could we hit the same problem there?

Reply:
In theory the second call to `Terminate` cannot hang because there is no active sync anymore (`doneCh == nil` means we are not syncing). I've made that even clearer now by removing the deferred call.